[Feat] Bridge Manager Rebase

TuDo1403 opened the pull request:
### Description
This PR rebase brach `dev` into branch `feat/bridge-admin`
Flowchart: https://www.notion.so/Bridge-Manager-Interaction-Flow-a51619f3e31549e78b6b8630ba79df74?pvs=4
### Contract changes

The table below shows the following info:
- **Logic**: the logic is changed.
- **ABI**: the ABI is changed.
- **Init data**: new storage field is declared and needs initializing.
- **Dependent**: needs to be changed due to changes in other contracts.

| **Contract name** | **Logic** | **ABI** | **Init data** | **Dependent** |
|-------------------|:---------:|:-------:|:-------------:|:-------------:|
| BridgeTracking    |    x       |     x    |        x       |               |
| GovernanceAdmin   |   x        |     x    |       x        |       x        |
| Maintenance       |           |         |               |               |
| SlashIndicator    |           |         |               |               |
| Staking           |           |         |               |               |
| StakingVesting    |           |         |               |               |
| ValidatorSet      |     x      |         |               |               |

### Checklist
- [x] I have clearly commented on all the main functions following the [NatSpec Format](https://docs.soliditylang.org/en/v0.8.0/natspec-format.html)
- [x] The box that allows repo maintainers to update this PR is checked
- [x] I tested locally to make sure this feature/fix works

Commit:
clmcc

Commit:
feat: remove bridge logic & comment test

Commit:
Merge branch 'dev' into feat/bridge-removal

Commit:
Update contracts/ronin/validator/RoninValidatorSet.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
featL remove bridge operator in applyValidatorCandidate

Commit:
format: named args

Commit:
feat: remove bridge reward deprecated at period

Commit:
feat: remove ValidatorFlag from getValidators()

Commit:
feat: remove isValidator & deprecate bridge enum

Commit:
fix: fix test

Commit:
Update contracts/ronin/staking/CandidateStaking.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
format: skip test

Commit:
format: add comments

Commit:
feat: update bridge admin contract

Commit:
Merge branch 'feat/bridge-removal' into feat/bridge-admin

Commit:
feat: add more checks

Commit:
feat: allow secondary address to update bridge operators

Commit:
feat: auth accounts

Commit:
fix: downgrade openzeppelin packages

Commit:
feat: getBridgeOperatorOf

Commit:
feat: bridge proposal

Commit:
feat: bridge voter weight

Commit:
feat: update domain separator

Commit:
format: minor refactor

Commit:
fix: fix logic

Commit:
feat: add minor test cases

Commit:
feat: add more tests

Commit:
feat: update tests

Commit:
feat: main chain bridge admin

Commit:
format: delete unused import

Commit:
format: rename function

Commit:
feat: remove bridge logic

Commit:
format: rename variable

Commit:
format: format code

Commit:
feat: update main chain bridge admin logic

Commit:
format: fix convention

Commit:
Update contracts/extensions/bridge-operator-governance/BridgeAdminOperator.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
format: refactor

Commit:
format: add doc for storage declarations

Commit:
feat: add ErrInvalidArguments

Commit:
format: rename test cases

Commit:
format: add variable visibility

nxqbao committed:
remove bridge in GA deploy

nxqbao committed:
rename contract

nxqbao committed:
add deploy script

Commit:
feat: update gas snapshot

Commit:
feat: update events

nxqbao committed:
add test setup

Commit:
fix: only Self Call instead of only Bridge contract

Commit:
feat: update events

Commit:
feat: add bridgeOperators when create contract

nxqbao committed:
rename to Bridge Manager

Commit:
fix: remove admin address

Commit:
fix: add propose function for bridge manager

Commit:
format: format code

nxqbao committed:
feat: support global vote for ronin bridge

Commit:
Update contracts/interfaces/IBridgeManager.sol

Co-authored-by: Bao <quocbao300@gmail.com>

Commit:
feat: check non duplicate between governor and bridge operator

nxqbao committed:
fix: remove check empty

nxqbao committed:
fix: init test

nxqbao committed:
fix: redundant nonDuplicate

nxqbao committed:
chore: remove console log

nxqbao committed:
test: fix configuration test

Commit:
feat: getProposalExpiryDuration

Commit:
feat: remove trusted orgs weight from ronin gateway

nxqbao committed:
test: restructure test set-type helper

nxqbao committed:
fix: constructor param order

nxqbao committed:
test: fix gateway pause enforcer test

Commit:
feat: remove global proposal

nxqbao committed:
fix: fix propose in test

nxqbao committed:
fix: fix config test

nxqbao committed:
fix: trusted vote count in test

nxqbao committed:
fix: fix other tests of gateway

nxqbao committed:
test: skip bridge set in GA test

Commit:
feat: split global proposal logics

nxqbao committed:
test: fix high nonce

Commit:
fix: readd expiryDuration to RoninGA

nxqbao committed:
test: fix mainchain GA test, add TODO

nxqbao committed:
test: fix integration bridge tracking test

Commit:
feat: TUint256 for custom slot uint256

nxqbao committed:
feat: add BridgeReward contract

nxqbao committed:
feat: check operator address must be payable

nxqbao committed:
feat: add initV3 for bridgeTracking

nxqbao committed:
fix: handle scattering reward twice

nxqbao committed:
fix: deployment scripts

nxqbao committed:
fix: remove expiry mainchain, fix deploy script

Commit:
feat: add SlashBridgeIndicator

Commit:
format: add comments, minor refactor

Commit:
feat: move onlySelfCall logic to `IdentityGuard`

nxqbao committed:
fix: remove mainchainGA deploy, add bridgeAdmin deploy

Commit:
fix: use `RONTransferHelper` for native transfer operation

Commit:
format: rename bridge slash contract

Commit:
format: rename `BridgeOperatorStatus` to `BridgeSlashInfo`

Commit:
feat: disable initializer

nxqbao committed:
refactor: re-split global GA, fix fixture

nxqbao committed:
fix: bridge tracking test setup

nxqbao committed:
fix: move query operator's info to tracking contract

nxqbao committed:
fix: rename 'penalize' to 'penalty' to fix grammar

nxqbao committed:
feat: integrate reward to slash contract

Commit:
fix: fix slash logic

Commit:
feat: remove auto remove functionality

Commit:
fix: fix event logic

nxqbao committed:
fix: order of assign to avoid misuse

nxqbao committed:
fix: fixture for hardhat test

Commit:
format: update comment for `execSlashBridgeOperators`

Commit:
fix: fix slashing logic

Commit:
feat: add common error `ErrInvalidReturnData`

nxqbao committed:
fix: handle zero vote in bridge reward

Commit:
feat: minor refactor

Commit:
format: add doc for `BridgeManager`

Commit:
fix: fix test

Commit:
format: remove duplicate `ErrZeroCodeContract` definition

Commit:
fix: fix test

nxqbao committed:
fix: merge code issue, add topup amount to test, fix test

nxqbao committed:
chore: lint coding convention

nxqbao committed:
chore: remove console log

nxqbao committed:
fix: fix wrong side assignment

nxqbao committed:
feat: replace getRoles by array to by enum

nxqbao committed:
fix: shadow declaration

nxqbao committed:
fix: begin block.timestamp in test

nxqbao committed:
feat: update version domain separator

Commit:
feat: add BridgeSlash test

Commit:
format: convention

Commit:
feat: check payable address in `updateBridgeOperator`

Commit:
feat: add `NewBridgeOperatorsAdded` event, `getAddedPeriodOf`

Commit:
test: add `test_bridgeSlash_recordEvents_onBridgeOperatorsAdded`

Commit:
feat: add `callbackRegisters` to `BridgeManager` constructor

Commit:
feat: add `supportsInterface`

Commit:
format: rename `BridgeManagerCallback` to `BridgeManagerCallbackRegister`

nxqbao committed:
fix: modify assertion, update docs

nxqbao committed:
feat: add receiveRON

Commit:
feat: add `test_slashTierLogic`

Commit:
fix: fix bridge manager deployments

nxqbao committed:
feat: add signature's method for bridge gw, fix test

nxqbao committed:
fix: fix method visibility

Commit:
format: remove magic numbers

nxqbao committed:
fix: default expiry duration on mainchain

nxqbao committed:
refactor: bring getSig and voted method into abstract contract

nxqbao committed:
fix: fix BridgeAdmin test

Commit:
format: refactor `BridgeSlash`

Commit:
test: add more tests

nxqbao committed:
feat: add local network proposal for BridgeManager

nxqbao committed:
fix: remove requireHasCode in constructor

nxqbao committed:
fix: fix integration test setup

Commit:
feat: add fuzz test for reward splitting logic `BridgeReward`

Commit:
test: add `test_WhenTotalBallotsZero_NotValidBridgeTrackingResponse`

Commit:
format: move `_requireSupportsInterface` to `IdentityGuard`

Commit:
feat: update `test_ExcludeNewlyAddedOperators_ExecSlashBridgeOperators`

nxqbao committed:
fix: handle share reward equally

nxqbao committed:
feat: differentiate share equally vs invalid tracking

Commit:
format: rename `test_SlashTierLogic` to `test_Fuzz_SlashTierLogic`

Commit:
format: add comments for `BridgeSlashTest` and `BridgeRewardTest`

Commit:
format: minor refactor

Commit:
feat: update `gas-snapshot` and `code size` logs

nxqbao committed:
fix: missing initializer in BridgeReward

Commit:
fix: load single field in `_getWeight`

Commit:
fix: remove error handling when callback revert

Commit:
format: remove unused lib

Commit:
feat: minor `extend` gas optimization

Commit:
Update contracts/interfaces/bridge/IBridgeManager.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

Commit:
format: remove `ChainTypeConsumer`, `AccessControlEnumerable`

Commit:
format: remove `Strings` lib

nxqbao committed:
Merge branch `dev` into feat/bridge-admin

nxqbao committed:
Merge branch 'dev' into feat/bridge-admin

ducthotran2010 commented on contracts/extensions/GovernanceAdmin.sol:
_Mark it as solved if my understanding abt the change in this file is correct:_

- Mainly added `IdentityGuard` lib to this contract & clean up duplicated methods
- Changed domain separator to avoid relay-attack

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,464 @@
+    address[] memory governors,
+    address[] memory bridgeOperators
+  ) internal nonDuplicate(governors.extend(bridgeOperators)) returns (bool[] memory addeds) {
+    uint256 length = bridgeOperators.length;
+    if (!(length == voteWeights.length && length == governors.length)) revert ErrLengthMismatch(msg.sig);
```
Please handle the case when length is equal to 0

TuDo1403 replied:
will skip add/remove operations when inputs are empty.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,464 @@
+          // get rid of stack too deep
+          // bridgeOperatorInfo.voteWeight = voteWeights[i].toUint96();
+          // accumulatedWeight += bridgeOperatorInfo.voteWeight
+          accumulatedWeight += bridgeOperatorInfo.voteWeight = voteWeights[i].toUint96();
```
Should `bridgeOperatorInfo.voteWeight` be a uint256 variable to avoid casting and SafeCast from OZ?

TuDo1403 replied:
struct `BridgeOperatorInfo` is packed into 1 slot to save gas.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,464 @@
+    uint256[] memory voteWeights,
+    address[] memory governors,
+    address[] memory bridgeOperators
+  ) internal nonDuplicate(governors.extend(bridgeOperators)) returns (bool[] memory addeds) {
```
Let's assume we have a set of 4 addresses: {A, B, C, D}, and 2 added bridge operators (BO) as follows:
BO 1: { govenor: A, bridgeOperator: B }
BO 2: { govenor: C, bridgeOperator: D }

Adding new BO(s) with reversed fields as following will be allowed?
BO 3: { govenor: B, bridgeOperator: A }
BO 4: { govenor: D, bridgeOperator: C }

TuDo1403 replied:
solution:
```solidity
addeds[i] = !(_governorSet.contains(governor) &&
_governorSet.contains(bridgeOperator) &&
_bridgeOperatorSet.contains(governor) &&
_bridgeOperatorSet.contains(bridgeOperator));
```

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,464 @@
+        if (addeds[i]) {
+          _governorSet.add(governor);
+
+          if (voteWeights[i].toUint96() == 0) revert ErrInvalidVoteWeight(msg.sig);
```
Check this from the start of for loop to avoid gas wasting

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,464 @@
+
+        addeds[i] = bridgeOperatorSet.add(bridgeOperator);
+
+        if (addeds[i]) {
```
Does it mean if the bridge operator is added before, this method will implicitly skip it?

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,464 @@
+
+    _notifyRegisters(IBridgeManagerCallback.onBridgeOperatorsAdded.selector, abi.encode(bridgeOperators, addeds));
+
+    emit BridgeOperatorsAdded(addeds, voteWeights, governors, bridgeOperators);
```
Does it mean when the value in `added` is false, the other fields with the same index don't take effect?

Commit:
fix: fix bug `BridgeManager:updateBridgeOperator`

Commit:
fix: fix `BridgeManager:_addBridgeOperators` and `BridgeManager:_removeBridgeOperators`

ducthotran2010 reviewed (changes requested):
Review checklist:
- [x] contracts/extensions/GovernanceAdmin.sol
- [ ] contracts/extensions/bridge-operator-governance/BridgeManager.sol
  -  [x] add methods
  -  [ ] others
- [ ] others

Commit:
feat: minor gas optimization

ducthotran2010 commented on contracts/utils/IdentityGuard.sol:
```diff
@@ -0,0 +1,82 @@
+   * @param addr The address to check.
+   */
+  function _requirePayableAddress(address addr) internal {
+    if (!_unsafeSendRON(payable(addr), 0, 0)) {
+      revert ErrNonpayableAddress(addr);
+    }
```
Please write some unit test to ensure it works as expected. Looks like this only works with EOA account, but the contract account with receive method will be reverted.

ducthotran2010 replied:
Correct me if I'm wrong. This instruction triggers [opCall](https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/instructions.go#L648) and jumps into [EVM CALL](https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/evm.go#L208): https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/evm.go#L275-L290

- The EOA accounts can bypass because the gas doesn't change
- The contract accounts will be reverted with the error code "out of gas" because the EVM triggers the receive method in the contract and [costs an operation gas](https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/interpreter.go#L226-L228)

ducthotran2010 replied:
Please note that because the call value is 0, the gas won't have stipend gas (2300): https://github.com/axieinfinity/ronin/blob/c0d708a7806d1eaec4ae98922bf8eb35375ac96d/core/vm/instructions.go#L666-L669

TuDo1403 replied:
`IdentityGuard:_requirePayableAddress` is removed since the guard is ineffective.

Commit:
feat: only emit `Notified` if contains registers

Commit:
fix: fix `BridgeSlash:execSlashBridgeOperators` use `totalVotes` as denominator instead of `totalBallots`,
feat: handle callback `IBridgeManagerCallback:onBridgeOperatorUpdated` to update slash info to new bridge operator

Commit:
feat: remove `IdentityGuard:_requirePayableAddress`

nxqbao committed:
fix: auto and manually sync reward

nxqbao committed:
fix: comparison in sync reward

Commit:
feat: `BridgeManager:_addBridgeOperators`, `BridgeManager:_removeBridgeOperators` skip add/remove operations when inputs are empty.

nxqbao committed:
fix: handle call reward

nxqbao committed:
fix: fix deployment

ducthotran2010 commented on contracts/ronin/gateway/RoninGatewayV2.sol:
```diff
@@ -502,14 +492,7 @@ contract RoninGatewayV2 is
-   */
-  function _getTotalTrustedWeight() internal view virtual returns (uint256) {
-    return IRoninTrustedOrganization(getContract(ContractType.RONIN_TRUSTED_ORGANIZATION)).countTrustedOrganizations();
+    return IBridgeManager(getContract(ContractType.BRIDGE_MANAGER)).totalBridgeOperators();
```
It should be `getTotalWeights` or sth like that instead
```suggestion
    return IBridgeManager(getContract(ContractType.BRIDGE_MANAGER)).getTotalWeights();
```

ducthotran2010 commented on contracts/ronin/gateway/RoninGatewayV2.sol:
```diff
@@ -538,24 +520,16 @@ contract RoninGatewayV2 is
+    unchecked {
+      for (uint _i; _i < _bridgeOperators.length; ++_i) {
+        if (_v.voteHashOf[_bridgeOperators[_i]] == _hash) {
+          _totalWeight++;
```
Since we have the weight for each bridge operator, why don't we use it instead of counting like this?

TuDo1403 replied:
make sense.

TuDo1403 replied:
what if `_v.voteHashOf[_bridgeOperators[_i]] != _hash`

ducthotran2010 commented on contracts/interfaces/bridge/IBridgeManager.sol:
```diff
@@ -0,0 +1,130 @@
+  /**
+   * @dev Returns total weights of the bridge voter list.
+   */
+  function getSumBridgeVoterWeights(address[] calldata governors) external view returns (uint256 sum);
```
```suggestion
  function sumGovernorWeights(address[] calldata governors) external view returns (uint256 sum);
```
We don't have bridge voter in this context

TuDo1403 replied:
fixed.

ducthotran2010 commented on contracts/interfaces/bridge/IBridgeManager.sol:
```diff
@@ -0,0 +1,130 @@
+  /**
+   * @dev Returns the weight of a bridge voter.
+   */
+  function getBridgeVoterWeight(address governor) external view returns (uint256);
```
Same, please check other places

ducthotran2010 commented on contracts/interfaces/bridge/IBridgeManager.sol:
Could you please provide a method by which we can get the whole list of bridge operators and the corresponding governor address, and weight?

TuDo1403 replied:
updated `BridgeManager:getFullBridgeOperatorInfos`

Commit:
fix: only return true for status if both GVs and BOs are updated.
fix: not allow BOs { B } to add A, and GVs { A } to add B

ducthotran2010 commented on contracts/utils/IdentityGuard.sol:
```diff
@@ -0,0 +1,71 @@
+    if (arr.hasDuplicate()) revert AddressArrayUtils.ErrDuplicated(msg.sig);
+  }
+
+  function _requireSupportsInterface(address register, bytes4 interfaceId) internal view {
```
I think this method assumes the input address is an IERC165 contract? Please add comments and note some main points like that

ducthotran2010 commented on contracts/utils/IdentityGuard.sol:
```diff
@@ -0,0 +1,71 @@
+    if (arr.hasDuplicate()) revert AddressArrayUtils.ErrDuplicated(msg.sig);
+  }
+
+  function _requireSupportsInterface(address register, bytes4 interfaceId) internal view {
```
I guess it can be use for common purposes so we can rename `register` to `contractAddr` or sth like that
```suggestion
  function _requireSupportsInterface(address contractAddr, bytes4 interfaceId) internal view {
```

TuDo1403 replied:
renamed.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManagerCallbackRegister.sol:
```diff
@@ -0,0 +1,123 @@
+        }
+      }
+
+      emit Notified(registers, statuses, returnDatas);
```
Consider including your calldata(s) to support debugging purposes

ducthotran2010 replied:
Because the calldata is deterministic per transaction so not including them is also fine

ducthotran2010 replied:
But when you have any intention to notify many types in a single transaction, please add the calldata to the event

TuDo1403 replied:
updated.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManagerCallbackRegister.sol:
Please add a method to get all registers

TuDo1403 replied:
added.

ducthotran2010 commented on contracts/types/operations/LibTUint256.sol:
Remove this file

TuDo1403 replied:
removed.

Commit:
format: rename `BridgeVoter` to `Governor`
feat: add `BridgeManager:getFullBridgeOperatorInfos`
  function _getTotalWeight() internal view virtual override returns (uint256) {
fix: `RoninGatewayV2:_getVoteWeight`, `RoninGatewayV2:`
format: remove `LibTUint256.sol`

Commit:
feat: add `BridgeManagerCallbackRegister:getCallbackRegisters`

Commit:
feat: add `bytes callData` to `IBridgeManagerCallbackRegister:Notified` event

Commit:
format: rename `register` to `contractAddr`
format: remove return data in `LibUint256Slot:store`

Commit:
format: remove `BridgeManager:getGovernorWeight`

Commit:
feat: minor optimization `BridgeManager:updateBridgeOperator`

Commit:
fix: fix `RoninGatewayV2` test

Commit:
format: add comment for `IdentityGuard:_requireSupportsInterface`

Commit:
format: add comments and minor refactor

Commit:
feat: use`IdentityGuard:_requireCreatedEOA` instead of `IdentityGuard:_requireNonZeroAddress`

Commit:
refactor: minor optimization

ducthotran2010 commented on contracts/extensions/RONTransferHelper.sol:
```diff
@@ -39,14 +39,14 @@ abstract contract RONTransferHelper {
    */
-  function _unsafeSendRON(address payable _recipient, uint256 _amount, uint256 _gas) internal returns (bool _success) {
-    (_success, ) = _recipient.call{ value: _amount, gas: _gas }("");
+  function _unsafeSendRON(address payable recipient, uint256 amount, uint256 gas) internal returns (bool success) {
```
Please consider renaming this method to avoid unintentionally using it instead of the above method

TuDo1403 replied:
renamed.

Commit:
format: minor refactor

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,499 @@
+          _requireCreatedEOA(bridgeOperator);
+          if (voteWeights[i].toUint96() == 0) revert ErrInvalidVoteWeight(msg.sig);
+
+          addeds[i] = !(_governorSet.contains(governor) &&
+            _governorSet.contains(bridgeOperator) &&
+            _bridgeOperatorSet.contains(governor) &&
+            _bridgeOperatorSet.contains(bridgeOperator));
```
```suggestion
          addeds[i] = !(_governorSet.contains(governor) ||
            _governorSet.contains(bridgeOperator) ||
            _bridgeOperatorSet.contains(governor) ||
            _bridgeOperatorSet.contains(bridgeOperator));
```

Should we create another mapping to store all addresses of governors and bridge operators and check in that mapping instead?

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,499 @@
+    if (!(length == voteWeights.length && length == governors.length)) revert ErrLengthMismatch(msg.sig);
+    addeds = new bool[](length);
+    // simply skip add operations if inputs are empty.
+    if (length != 0) {
```
```suggestion
    if (length == 0) return adds;

```

Commit:
format: rename `getSumGovernorWeights` to `sumGovernorsWeight`

Commit:
fix: fix `_addBridgeOperators` branching logic

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+    EnumerableSet.AddressSet storage _bridgeOperatorSet = _getBridgeOperatorSet();
+    mapping(address => BridgeOperatorInfo) storage _governorToBridgeOperatorInfo = _getGovernorToBridgeOperatorInfo();
+
+    // get rid of stack too deep
```
```suggestion
    // avoid stack too deep
```

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+          _governorOf[bridgeOperator] = governor;
+          // get rid of stack too deep
+          // bridgeOperatorInfo.voteWeight = voteWeights[i].toUint96();
+          // accumulatedWeight += bridgeOperatorInfo.voteWeight
```
I don't get error stack too deep

TuDo1403 replied:
removed.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+    // simply skip add operations if inputs are empty.
+    if (length == 0) {
+      return addeds;
+    }
```
```suggestion

    if (length == 0) return addeds;
```

Commit:
fix: fix test

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+    // return false if currentBridgeOperator unexists in _bridgeOperatorSet
+    if (!_bridgeOperatorSet.remove(currentBridgeOperator)) {
+      revert ErrUnauthorized(msg.sig, RoleAccess.GOVERNOR);
+    }
```
This method can grant any caller to be a governor if the method `add` is handled mistakenly for null address `0x0`. Consider checking the existence of `msg.sender` before.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+    if (!updated) revert ErrBridgeOperatorAlreadyExisted(newBridgeOperator);
+
+    _gorvernorToBridgeOperatorInfo[msg.sender].addr = newBridgeOperator;
+    _governorOf[newBridgeOperator] = msg.sender;
+
+    delete _governorOf[currentBridgeOperator];
```
If `currentBridgeOperator==newBridgeOperator`, we may lose the ability to get the governor from that bridge operator address

TuDo1403 replied:
please review the below solution:
```solidity
/**
   * @inheritdoc IBridgeManager
   */
  function updateBridgeOperator(address newBridgeOperator) external returns (bool updated) {
    _requireCreatedEOA(newBridgeOperator);

    mapping(address => address) storage _governorOf = _getGovernorOf();
    EnumerableSet.AddressSet storage _governorSet = _getGovernorsSet();
    EnumerableSet.AddressSet storage _bridgeOperatorSet = _getBridgeOperatorSet();
    mapping(address => BridgeOperatorInfo) storage _gorvernorToBridgeOperatorInfo = _getGovernorToBridgeOperatorInfo();

    address currentBridgeOperator = _gorvernorToBridgeOperatorInfo[msg.sender].addr;

    if (currentBridgeOperator == newBridgeOperator) {
      revert ErrBridgeOperatorAlreadyExisted(newBridgeOperator);
    }
    // return false if `governor` unexists in `_governorSet`
    // return false if `currentBridgeOperator` unexists in `_bridgeOperatorSet`
    if (!(_governorSet.contains(msg.sender) && _bridgeOperatorSet.contains(currentBridgeOperator))) {
      revert ErrUnauthorized(msg.sig, RoleAccess.GOVERNOR);
    }

    updated = _bridgeOperatorSet.remove(currentBridgeOperator) && _bridgeOperatorSet.add(newBridgeOperator);
    if (!updated) revert ErrBridgeOperatorUpdateFailed(newBridgeOperator);

    _gorvernorToBridgeOperatorInfo[msg.sender].addr = newBridgeOperator;
    _governorOf[newBridgeOperator] = msg.sender;

    delete _governorOf[currentBridgeOperator];

    _notifyRegisters(
      IBridgeManagerCallback.onBridgeOperatorUpdated.selector,
      abi.encode(currentBridgeOperator, newBridgeOperator, updated)
    );

    emit BridgeOperatorUpdated(msg.sender, currentBridgeOperator, newBridgeOperator);
  }
```

ducthotran2010 replied:
Use this one:
```solidity
  function updateBridgeOperator(address newBridgeOperator) external returns (bool updated) {
    _requireCreatedEOA(newBridgeOperator);

    // Checks authorization
    EnumerableSet.AddressSet storage _governorSet = _getGovernorsSet();
    if (!_governorSet.contains(msg.sender)) revert ErrUnauthorized(msg.sig, RoleAccess.GOVERNOR);

    // Queries the previous bridge operator
    mapping(address => BridgeOperatorInfo) storage _gorvernorToBridgeOperatorInfo = _getGovernorToBridgeOperatorInfo();
    address currentBridgeOperator = _gorvernorToBridgeOperatorInfo[msg.sender].addr;
    if (currentBridgeOperator == newBridgeOperator) {
      revert ErrBridgeOperatorAlreadyExisted(newBridgeOperator);
    }

    // Tries replace the bridge operator
    EnumerableSet.AddressSet storage _bridgeOperatorSet = _getBridgeOperatorSet();
    updated = _bridgeOperatorSet.remove(currentBridgeOperator) && _bridgeOperatorSet.add(newBridgeOperator);
    if (!updated) return updated;

    mapping(address => address) storage _governorOf = _getGovernorOf();
    delete _governorOf[currentBridgeOperator];
    _governorOf[newBridgeOperator] = msg.sender;
    _gorvernorToBridgeOperatorInfo[msg.sender].addr = newBridgeOperator;

    _notifyRegisters(
      IBridgeManagerCallback.onBridgeOperatorUpdated.selector,
      abi.encode(currentBridgeOperator, newBridgeOperator, updated)
    );
    emit BridgeOperatorUpdated(msg.sender, currentBridgeOperator, newBridgeOperator);
  }
```

ducthotran2010 replied:
Later if you decide to move the modifier `onlyGovernor` to this file, consider removing the "checks authorization" step

TuDo1403 replied:
this line would corrupt storage state:
```solidity
updated = _bridgeOperatorSet.remove(currentBridgeOperator) && _bridgeOperatorSet.add(newBridgeOperator);
```

Commit:
feat: format `voteWeights` to `uint96`

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+    address governor;
+    address bridgeOperator;
+    uint256 accumulatedWeight;
+    BridgeOperatorInfo memory bridgeOperatorInfo;
```
I think turning it to be `storage` will reduce gas cost
```suggestion
    BridgeOperatorInfo store bridgeOperatorInfo;
```

TuDo1403 replied:
declaring storage in for loop is will trigger unnecessary `sload` operation which is expected to cost more gas.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+
+  function _getBridgeOperators() internal view returns (address[] memory) {
+    return _getBridgeOperatorSet().values();
+  }
+
+  function _getGovernors() internal view returns (address[] memory) {
+    return _getGovernorsSet().values();
+  }
```
Why should we create new methods just for getting inner values?

TuDo1403 replied:
we prefer using `internal` and `external` methods over using `public` methods.

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,503 @@
+  /**
+   * @inheritdoc IBridgeManager
+   */
+  function getBridgeOperatorOf(address[] calldata governors) external view returns (address[] memory bridgeOperators_) {
+    uint256 length = governors.length;
+    bridgeOperators_ = new address[](length);
+
+    mapping(address => BridgeOperatorInfo) storage _gorvernorToBridgeOperator = _getGovernorToBridgeOperatorInfo();
+    for (uint256 i; i < length; ) {
+      bridgeOperators_[i] = _gorvernorToBridgeOperator[governors[i]].addr;
+      unchecked {
+        ++i;
+      }
+    }
+  }
+
+  function getGovernorsOf(address[] calldata bridgeOperators) external view returns (address[] memory governors) {
+    uint256 length = bridgeOperators.length;
+    governors = new address[](length);
+    mapping(address => address) storage _governorOf = _getGovernorOf();
+
+    for (uint256 i; i < length; ) {
+      governors[i] = _governorOf[bridgeOperators[i]];
+      unchecked {
+        ++i;
+      }
+    }
+  }
```
I cannot see any loc that uses these methods

TuDo1403 replied:
I think it maybe useful for future usage.

ducthotran2010 commented on contracts/utils/IdentityGuard.sol:
```diff
@@ -0,0 +1,91 @@
+   * This internal function is used to ensure that the provided address is a valid externally owned account (EOA).
+   * It checks the codehash of the address against a predefined constant to confirm that the address is a created EOA.
+   */
+  function _requireCreatedEOA(address addr) internal view {
+    bytes32 codehash = addr.codehash;
+    if (codehash != CREATED_ACCOUNT_HASH) revert ErrAddressIsNotCreatedEOA(addr, codehash);
+  }
```
Note: This method only works with non-state EOA accounts

TuDo1403 replied:
commented.

Commit:
Update contracts/extensions/bridge-operator-governance/BridgeManager.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

ducthotran2010 commented on contracts/utils/IdentityGuard.sol:
```diff
@@ -0,0 +1,91 @@
+   * @dev Internal function to require that the specified contract supports the given interface.
+   * @param contractAddr The address of the contract to check for interface support.
+   * @param interfaceId The interface ID to check for support.
+   * @dev If the contract does not support the interface, a revert with the corresponding error message is triggered.
```
Do not use more than one `@dev`
```suggestion
   * @notice If the contract does not support the interface `interfaceId` or EIP165, a revert with the corresponding error message is triggered.
```

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManagerCallbackRegister.sol:
```diff
@@ -0,0 +1,133 @@
+    unregistereds = _unregisterCallbacks(registers);
+  }
+
+  function getCallbackRegisters() external view returns (address[] memory registers) {
```
```suggestion
  /**
   * @inheritdoc IBridgeManagerCallbackRegister
   */
  function getCallbackRegisters() external view returns (address[] memory registers) {
```

Commit:
Update contracts/utils/IdentityGuard.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

ducthotran2010 commented on contracts/ronin/gateway/RoninBridgeManager.sol:
```diff
@@ -0,0 +1,243 @@
+
+  function _getWeight(address _governor) internal view virtual override returns (uint256) {
+    return _getGovernorToBridgeOperatorInfo()[_governor].voteWeight;
+  }
```
Please add comment

TuDo1403 replied:
added.

ducthotran2010 commented on contracts/ronin/gateway/RoninBridgeManager.sol:
```diff
@@ -0,0 +1,243 @@
+  modifier onlyGovernor() {
+    _requireGovernor();
+    _;
+  }
```
Both `<Ronin|Mainchain>BridgeManager` use this modifer

TuDo1403 replied:
moved to base contract `BridgeManager`

ducthotran2010 commented on contracts/mainchain/MainchainBridgeManager.sol:
```diff
@@ -0,0 +1,103 @@
+  uint256 private constant DEFAULT_EXPIRY_DURATION = 1 << 255;
+
+  modifier onlyGovernor() {
+    _requireGorvernor();
```
```suggestion
    _requireGovernor();
```

TuDo1403 replied:
fixed.

ducthotran2010 commented on contracts/mainchain/MainchainBridgeManager.sol:
```diff
@@ -0,0 +1,103 @@
+    });
+  }
+
+  function _getMinimumVoteWeight() internal view override returns (uint256) {
+    return minimumVoteWeight();
+  }
+
+  /**
+   * @dev Returns the expiry duration for a new proposal.
+   */
+  function getProposalExpiryDuration() external view returns (uint256) {
+    return _getProposalExpiryDuration();
+  }
+
+  function _getTotalWeights() internal view override returns (uint256) {
+    return getTotalWeights();
+  }
+
+  function _requireGorvernor() private view {
+    if (_getWeight(msg.sender) == 0) revert ErrUnauthorized(msg.sig, RoleAccess.GOVERNOR);
+  }
+
+  function _sumWeights(address[] memory governors) internal view override returns (uint256) {
+    return sumGovernorsWeight(governors);
+  }
+
+  function _getChainType() internal pure override returns (ChainType) {
+    return ChainType.Mainchain;
+  }
+
+  function _getWeight(address _governor) internal view virtual override returns (uint256) {
+    BridgeOperatorInfo memory bridgeOperatorInfo = _getGovernorToBridgeOperatorInfo()[_governor];
+    return bridgeOperatorInfo.voteWeight;
+  }
```
Add comments

TuDo1403 replied:
added.

ducthotran2010 commented on contracts/extensions/sequential-governance/governance-relay/CommonGovernanceRelay.sol:
```diff
@@ -83,61 +83,9 @@ abstract contract GovernanceRelay is CoreGovernance {
-      ECDSA.toTypedDataHash(_domainSeparator, Ballot.hash(_globalProposalHash, Ballot.VoteType.Against))
-    );
-  }
+  function _getWeight(address _governor) internal view virtual returns (uint256);
```
I dont see any use case of this method

TuDo1403 replied:
`_getWeight` is used in `CommonGovernanceProposal`, `GovernanceProposal`, `CommonGovernanceRelay`.

ducthotran2010 commented on contracts/mainchain/MainchainBridgeManager.sol:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  function _getWeight(address _governor) internal view virtual override returns (uint256) {
+    BridgeOperatorInfo memory bridgeOperatorInfo = _getGovernorToBridgeOperatorInfo()[_governor];
+    return bridgeOperatorInfo.voteWeight;
```
Don't load the whole struct to memory just for retrieving a field

TuDo1403 replied:
fixed.

ducthotran2010 commented on contracts/interfaces/IRoninBridgeManager.sol:
Looks like the `IRoninBridgeManager` and `IMainchainBridgeManager` share the same interface. And I cannot see any use case of them.

TuDo1403 replied:
removed.

ducthotran2010 reviewed (changes requested):
Looks like this PR miss handling Mainchain Bridge, I cannot see any change in `MainchainGatewayV2`

Commit:
fix: fix duplicate address update in `BridgeManager:updateBridgeOperator`
format: remove stack too deep comment
format: remove unnecessary underscore prefix

Commit:
feat: move `onlyGovernor` to `BridgeManager`

Commit:
feat: add `memory-safe` assembly
format: minor refactor

Commit:
format: remove return value `updated` for `BridgeManager:updateBridgeOperator`
feat: add view methods `getBridgeOperatorWeights`, `getBridgeOperatorWeight`
feat: update `MainchainGatewayV2`
format: add comments for internal methods

Commit:
fix: reduce risk of reentrancy

Commit:
feat: minor refactor

Commit:
feat: add `MainchainGatewayV2:supportsInterface`

Commit:
format: minor refactor

Commit:
feat: remove unused functions

Commit:
format: rename `_unsafeSendRON` to `_unsafeSendRONLimitGas`

Commit:
format: deprecate storage var

ducthotran2010 commented on contracts/extensions/bridge-operator-governance/BridgeManager.sol:
```diff
@@ -0,0 +1,580 @@
+  }
+
+  /**
+   * @inheritdoc IBridgeManager
```
```suggestion
   * @inheritdoc IBridgeManager
   * @notice This method checks authorization by querying the corresponding operator of the msg.sender and then
   * attempts to remove it from the `_bridgeOperatorSet` for gas optimization. In case we allow a governor can leave
   * their operator address blank null `address(0)`, consider add authorization check.
```

Commit:
Update contracts/extensions/bridge-operator-governance/BridgeManager.sol

Co-authored-by: Duc Tho Tran <ducthotran2010@gmail.com>

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
Add comments

TuDo1403 replied:
added.

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
```diff
@@ -0,0 +1,256 @@
+    uint256 period
+  ) external onlyContract(ContractType.BRIDGE_TRACKING) {
+    // Only sync the period that is after the latest rewarded period.
+    if (period != _latestRewardedPeriod + 1) {
+      // Emit event instead of revert since bridge tracking and voting process depends on this.
+      emit BridgeRewardSyncTooFarPeriod(period, _latestRewardedPeriod);
+      return;
+    }
```
if a period is missed this contract will never distribute reward?

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
```diff
@@ -0,0 +1,256 @@
+    IBridgeTracking bridgeTrackingContract = IBridgeTracking(getContract(ContractType.BRIDGE_TRACKING));
+
+    uint256 currentPeriod = IRoninValidatorSet(getContract(ContractType.VALIDATOR)).currentPeriod();
+    if (currentPeriod <= _latestRewardedPeriod) {
+      revert ErrPeriodAlreadyProcessed(currentPeriod, _latestRewardedPeriod);
+    }
```
Since this method is called by the bridge service automatically, should we simply return the method instead of reverting

TuDo1403 replied:
applied.

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
```diff
@@ -0,0 +1,256 @@
+      revert ErrUnauthorizedCall(msg.sig);
+    }
+
+    IBridgeManager bridgeManagerContract = IBridgeManager(getContract(ContractType.BRIDGE_MANAGER));
+    IBridgeTracking bridgeTrackingContract = IBridgeTracking(getContract(ContractType.BRIDGE_TRACKING));
```
Move these LoC to be after the period check

TuDo1403 replied:
moved.

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
```diff
@@ -0,0 +1,256 @@
+    BridgeRewardInfo storage _iRewardInfo = _rewardInfo[operator];
+    if (isSlashed) {
+      _iRewardInfo.slashed += reward;
+      emit BridgeRewardSlashed(operator, reward);
+    } else {
+      _iRewardInfo.claimed += reward;
+      if (_unsafeSendRONLimitGas({ recipient: payable(operator), amount: reward, gas: 0 })) {
+        emit BridgeRewardScattered(operator, reward);
+      } else {
+        emit BridgeRewardScatterFailed(operator, reward);
```
These 3 events should emitted with the period

TuDo1403 replied:
updated.

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
```diff
@@ -0,0 +1,256 @@
+  ) internal onlyContract(ContractType.BRIDGE_TRACKING) {
+    ++_latestRewardedPeriod;
+
+    bool isSlashed;
```
`shouldSlash` or `slashed`

TuDo1403 replied:
renamed.

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
```diff
@@ -0,0 +1,256 @@
+    uint256[] memory slashedDurationList = _getSlashInfo(operators);
+
+    // Validate should share the reward equally
+    bool isSharingRewardEqually = _isSharingRewardEqually(totalBallot, totalVote, ballots);
```
`shouldShareEqually` or `sharedEqually`

TuDo1403 replied:
renamed.

ducthotran2010 commented on contracts/ronin/gateway/BridgeReward.sol:
```diff
@@ -0,0 +1,256 @@
+    uint256 numBridgeOperators = operators.length;
+
+    for (uint256 i; i < operators.length; ) {
+      (reward, isSlashed) = _calcRewardAndCheckSlashedStatus(
+        isSharingRewardEqually,
+        numBridgeOperators,
+        rewardPerPeriod,
+        ballots[i],
+        totalBallot,
+        period,
+        slashedDurationList[i]
+      );
```
Use `({ ... })` pattern

TuDo1403 replied:
applied.

Commit:
feat: minor refactor

Commit:
feat: move `event`, `struct`, `enum` to to other interface
format: rename variable
feat: allow sync reward for multiple periods
feat: add bridge tracking response validation for `BridgeSlash`
fix: remove `onlyContract` guard in `BridgeReward:_syncReward`
format: use named arguments
feat: custom slots for `BridgeReward`
format: add documentations

Commit:
format: add doc

Commit:
fix: remove `BridgeSlash:setContract`

Commit:
feat: only slash if `totalVotes` >= `MINIMUM_VOTE_THRESHOLD`

Commit:
fix: BridgeReward:LATEST_REWARDED_PERIOD_SLOT empty value when `syncReward`

Commit:
fix: fix mishandle `execSyncReward`
format: add comments
feat: add `getTotalRewardsToppedUp`, `getTotalRewardsScattered`
feat: prevent mis transfer RON to logic contract

Commit:
format: rename var

Commit:
feat: remove dead code

Commit:
format: minor refactor

Commit:
format: minor refactor

Commit:
feat: add `_requireNonZeroAddress` in `_requireCreatedEOA`

nxqbao committed:
fix: emit event

nxqbao committed:
chore: remove plural form of `total-` var

Commit:
fix: prevent reverting in `BridgeTracking:recordVote`

nxqbao committed:
chore: fix convention

nxqbao committed:
fix test

nxqbao committed:
fix: collided vars

Commit:
fix: revert when invalid`periodLength`

Commit:
feat: add `ErrSyncTooFarPeriod` if `period > latestRewardedPeriod + 1`

Commit:
fix: use `getBridgeOperatorOf` instead of `getBridgeOperators`

Commit:
feat: add fork test

Commit:
feat: allow `proxyAdmin` to query for `supportsInterface`

nxqbao committed:
chore: update comment, fix test method visibility

Commit:
ci: exclude fork test

Commit:
fix: exclude fork test

Commit:
fix: exclude fork test in ci

Commit:
fix: remove `_requireCreatedEOA`

nxqbao reviewed (approved):
Approved after deployment done.

ducthotran2010 reviewed (approved):
Too many changes in this PR. Should put a limitation on a single PR next time (eg, won't merge for PR with more than +-1000 loc)
